### DISCUSSION TITLE
Make GridPatchDataset cacheable by adding shapes parameter

### DISCUSSION THREAD

**Discussion by schellenchris** [Project-MONAI/MONAI#5662] [Ideas]
I noticed computing the slices each epoch isn't really fast so I made the GridPatchDataset cacheable by adding a shapes parameter.
I compute the shapes `shapes = [nib.load(img).shape for img in imgs]` in advance, it's pretty fast. I hand the shapes over to the GridPatchDataset to compute the length of the dataset. For `__getitem__ ` I store the sliced image so I just need to slice it once. For multi dimension slicing and overlaps it could get more complex but I think it's achievable. Handing the GridPatchDataset into CacheDataset managed to speed up my training.

**Comment by wyli** [comment]
do you mean `GridPatchDataset` to inherit `Dataset` in addition to `IterableDataset`? would be great to have a small example to briefly show the idea...

**Reply by schellenchris** [reply]
inheriting `Dataset` would be the clean solution I think. I just implemented `__len__` and `__getitem__` functions to make it work the quick way. I linked you my `GridPatchDataset`
https://github.com/schellenchris/MONAIFiles/blob/73f98e197be4ad318b3cb150d6c9270fc504e269/grid_dataset.py#L205-L236
